### PR TITLE
fix: tests

### PR DESCRIPTION
**Pull request by sattvikc**
## Summary of change

(A few sentences about this PR)

## Related issues

- Link to issue1 here
- Link to issue1 here

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your
changes work. Bonus points for screenshots and videos!)

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here
highlighting the necessary changes)

## Checklist for important updates

- [ ] Changelog has been updated
    - [ ] If there are any db schema changes, mention those changes clearly
- [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
- [ ] `pluginInterfaceSupported.json` file has been updated (if needed)
- [ ] Changes to the version if needed
    - In `build.gradle`
- [ ] If added a new paid feature, edit the `getPaidFeatureStats` function in FeatureFlag.java file
- [ ] Had installed and ran the pre-commit hook
- [ ] If there are new dependencies that have been added in `build.gradle`, please make sure to add them
  in `implementationDependencies.json`.
- [ ] Update function `getValidFields` in `io/supertokens/config/CoreConfig.java` if new aliases were added for any core config (similar to the `access_token_signing_key_update_interval` config alias).
- [ ] Issue this PR against the latest non released version branch.
    - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the
      latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR

- [ ] Item1
- [ ] Item2

### CODE REVIEW

File: src/main/java/io/supertokens/authRecipe/AuthRecipe.java [hunk 1]
```diff
@@ -840,7 +840,8 @@ private static void deleteUserHelper(TransactionConnection con, AppIdentifierWit
 
         if (!removeAllLinkedAccounts) {
             deleteAuthRecipeUser(con, appIdentifierWithStorage, userIdToDeleteForAuthRecipe,
-                    !userIdToDeleteForAuthRecipe.equals(userToDelete.getSupertokensUserId()));
+                    !userIdToDeleteForAuthRecipe.equals(userToDelete.getSupertokensUserId()) ||
+                            userToDelete.loginMethods.length == 1);
```
[the remark this answers is not in the document]
Comment (rishabhpoddar) [reply src/main/java/io/supertokens/authRecipe/AuthRecipe.java:844]: Instead, on line 819, check if primaryUserIdToDeleteNonAuthRecipe == null, and then if it is, call whats on line 854

File: src/main/java/io/supertokens/multitenancy/Multitenancy.java [hunk 2]
```diff
@@ -396,7 +396,7 @@ public static boolean addUserIdToTenant(Main main, TenantIdentifierWithStorage t
                 String tenantId = tenantIdentifierWithStorage.getTenantId();
                 AuthRecipeUserInfo userToAssociate = storage.getPrimaryUserById_Transaction(tenantIdentifierWithStorage.toAppIdentifier(), con, userId);
 
-                if (userToAssociate.isPrimaryUser) {
+                if (userToAssociate != null && userToAssociate.isPrimaryUser) {
                     Set<String> emails = new HashSet<>();
                     Set<String> phoneNumbers = new HashSet<>();
                     Set<LoginMethod.ThirdParty> thirdParties = new HashSet<>();
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/multitenancy/Multitenancy.java:402]: the checks below for same email etc for account linking should check:
- if the login method is the same, throw duplicaremalexception etc..
- if it's a different login method, then we shuold throw a new exception type -> UserTenantAssociationNotAllowed with a reason string -> if this is on the backend SDK, then we should have the status reflected in there too

Comment (rishabhpoddar) [reply src/main/java/io/supertokens/multitenancy/Multitenancy.java:402]: Should the other places where this type of check also happens consider the login method being different?

File: src/main/java/io/supertokens/webserver/api/multitenancy/DisassociateUserFromTenant.java [hunk 3]
```diff
@@ -79,7 +82,10 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws I
             JsonObject result = new JsonObject();
             result.addProperty("status", "UNKNOWN_USER_ID_ERROR");
             super.sendJsonResponse(200, result, resp);
-
+        } catch (DisassociationNotAllowedException e) {
+            JsonObject result = new JsonObject();
+            result.addProperty("status", "DISASSOCIATION_NOT_ALLOWED_ERROR");
+            super.sendJsonResponse(200, result, resp);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/multitenancy/DisassociateUserFromTenant.java:88]: add a reason string, and also have this showed on the backend sdk

File: src/main/java/io/supertokens/thirdparty/ThirdParty.java [hunk 4]
```diff
@@ -250,35 +247,75 @@ private static SignInUpResponse signInUpHelper(TenantIdentifierWithStorage tenan
...
-                    return null;
-                });
+                            storage.commitTransaction(con);
+                            return null;
+                        });
+                    }
+                }
 
                 AuthRecipeUserInfo user = getUser(tenantIdentifierWithStorage, thirdPartyId, thirdPartyUserId);
                 return new SignInUpResponse(false, user);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/thirdparty/ThirdParty.java:321]: the try catch below is incorrectly scoped

File: src/main/java/io/supertokens/thirdparty/ThirdParty.java [hunk 5]
```diff
@@ -216,12 +216,11 @@ private static SignInUpResponse signInUpHelper(TenantIdentifierWithStorage tenan
                 AuthRecipeSQLStorage authRecipeStorage =
                         (AuthRecipeSQLStorage) tenantIdentifierWithStorage.getAuthRecipeStorage();
 
-                storage.startTransaction(con -> {
+                { // Try without transaction
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/thirdparty/ThirdParty.java:219]: add more comment - why are we doing this?

File: src/main/java/io/supertokens/webserver/api/multitenancy/DisassociateUserFromTenant.java (hunk 3 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/multitenancy/DisassociateUserFromTenant.java:88]: Done

File: src/main/java/io/supertokens/thirdparty/ThirdParty.java (hunk 4 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/thirdparty/ThirdParty.java:321]: Done

File: src/main/java/io/supertokens/thirdparty/ThirdParty.java (hunk 5 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/thirdparty/ThirdParty.java:219]: Done

File: src/main/java/io/supertokens/authRecipe/AuthRecipe.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/authRecipe/AuthRecipe.java:844]: Done

**Pull request merged**